File: 02/Exercise1/Program.cs
using System;
using System.IO;
using System.Linq;

namespace Exercise1
{
    class Program
    {
        static void Main(string[] args)
        {
            var lines = File.ReadAllLines(args[0]);
            var validPasswordCount = 0;
            foreach (var line in lines) {
                var lineParts = line.Split(" ");
                var rangeParts = lineParts[0].Split("-");
                var min = int.Parse(rangeParts[0]);
                var max = int.Parse(rangeParts[1]);
                var character = lineParts[1][0];
                var password = lineParts[2];
                var characterCount = password.ToCharArray().Count(c => c == character);
                if (characterCount >= min && characterCount <= max) validPasswordCount++;
            }
            Console.WriteLine(validPasswordCount);
        }
    }
}

File: 02/fsharp/Exercise1/Program.fs
open System
open System.IO

type Policy = {
    Range: int * int;
    Letter: char;
}

let count x = Seq.filter ((=) x) >> Seq.length

let isValidPassword (policy, password) =
    let (min, max) = policy.Range
    password |> count policy.Letter |> (fun count -> count >= min && count <= max)

let parsePolicyAndPassword (text: string) =
    let parts = text.Split (" ")
    let [| min; max |] = parts.[0].Split ("-") |> Array.map int
    let letter = parts.[1] |> Seq.head
    let password = parts.[2]
    ({ Range = (min, max); Letter = letter }, password)

[<EntryPoint>]
let main argv =
    let input =
        File.ReadAllLines(argv.[0])
        |> Array.map parsePolicyAndPassword
    let validPasswords =
        input
        |> Array.filter isValidPassword
    Console.WriteLine(validPasswords.Length)
    0 // return an integer exit code
